### PR TITLE
[OpenMP] Add Environment Variable to disable Reuse of Blocks for High Loop Trip Counts

### PR DESCRIPTION
**Pull request by tgymnich**
Sometimes it might be beneficial to spawn more thread blocks instead of reusing existing for multiple loop iterations.

**Alternatives considered:**

Make `DefaultNumBlocks` settable via an environment variable.

### CODE REVIEW

File: openmp/libomptarget/plugins-nextgen/common/include/PluginInterface.h [hunk 1]
```diff
@@ -904,6 +910,9 @@ struct GenericDeviceTy : public DeviceAllocatorTy {
   UInt32Envar OMPX_MinThreadsForLowTripCount =
       UInt32Envar("LIBOMPTARGET_MIN_THREADS_FOR_LOW_TRIP_COUNT", 32);
 
+  BoolEnvar OMPX__ReuseBlocksForHighTripCount =
```
Comment (jdoerfert) [inline openmp/libomptarget/plugins-nextgen/common/include/PluginInterface.h:913]: One `_` is enough. We need a comment here explaining the meaning. Also need to add it to the docs.
Finally, can you add this into the test that checks for the trip count adjustment as a new run line with the variable set?

File: openmp/libomptarget/plugins-nextgen/common/include/PluginInterface.h [hunk 2]
```diff
@@ -829,6 +829,12 @@ struct GenericDeviceTy : public DeviceAllocatorTy {
     return OMPX_MinThreadsForLowTripCount;
   }
 
+  /// Whether or not to reuse blocks for high trip count loops.
+  /// @see OMPX__ReuseBlocksForHighTripCount
+  virtual bool getReuseBlocksForHighTripCount() {
```
Comment (jhuber6) [inline openmp/libomptarget/plugins-nextgen/common/include/PluginInterface.h:834]: Why is this virtual?

File: openmp/libomptarget/plugins-nextgen/common/src/PluginInterface.cpp [hunk 3]
```diff
@@ -705,8 +705,12 @@ uint64_t GenericKernelTy::getNumBlocks(GenericDeviceTy &GenericDevice,
       TripCountNumBlocks = LoopTripCount;
     }
   }
+
+  uint32_t PreferredNumBlocks = TripCountNumBlocks;
   // If the loops are long running we rather reuse blocks than spawn too many.
-  uint32_t PreferredNumBlocks = std::min(TripCountNumBlocks, DefaultNumBlocks);
+  if (GenericDevice.getReuseBlocksForHighTripCount()) {
+    PreferredNumBlocks = std::min(TripCountNumBlocks, DefaultNumBlocks);
+  }
```
Comment (jhuber6) [inline openmp/libomptarget/plugins-nextgen/common/src/PluginInterface.cpp:713]:
```suggestion
  if (GenericDevice.getReuseBlocksForHighTripCount())
    PreferredNumBlocks = std::min(TripCountNumBlocks, DefaultNumBlocks);
```

**Review by jhuber6 — COMMENTED** [review]
The sheer number of environment variables we seem to be accumulating is concerning. We really need an API function that toggles features on or off instead, but that's an entirely different discussion.

File: openmp/libomptarget/plugins-nextgen/common/include/PluginInterface.h (hunk 2 above)
Comment (tgymnich) [reply openmp/libomptarget/plugins-nextgen/common/include/PluginInterface.h]: fixed

File: openmp/libomptarget/plugins-nextgen/common/include/PluginInterface.h (hunk 1 above)
Comment (tgymnich) [reply openmp/libomptarget/plugins-nextgen/common/include/PluginInterface.h]: fixed

File: openmp/libomptarget/test/offloading/high_trip_count_block_limit.cpp [hunk 4]
```diff
@@ -0,0 +1,29 @@
+// RUN: %libomptarget-compilexx-generic && env
+// LIBOMPTARGET_REUSE_BLOCKS_FOR_HIGH_TRIP_COUNT=False %libomptarget-run-generic
+// 2>&1 | %fcheck-generic UNSUPPORTED: aarch64-unknown-linux-gnu UNSUPPORTED:
+// aarch64-unknown-linux-gnu-LTO UNSUPPORTED: x86_64-pc-linux-gnu UNSUPPORTED:
+// x86_64-pc-linux-gnu-LTO UNSUPPORTED: s390x-ibm-linux-gnu UNSUPPORTED:
+// s390x-ibm-linux-gnu-LTO
```
Comment (jdoerfert) [inline openmp/libomptarget/test/offloading/high_trip_count_block_limit.cpp:6]: This got formatted badly.

Also, add a second RUN line to test the default. Then don't check for pass but just print the values and CHECK for the two different expected ones.

File: openmp/docs/design/Runtimes.rst [hunk 5]
```diff
@@ -742,6 +742,7 @@ variables is defined below.
     * ``LIBOMPTARGET_JIT_PRE_OPT_IR_MODULE=<out:Filename> (LLVM-IR file)``
     * ``LIBOMPTARGET_JIT_POST_OPT_IR_MODULE=<out:Filename> (LLVM-IR file)``
     * ``LIBOMPTARGET_MIN_THREADS_FOR_LOW_TRIP_COUNT=<Num> (default: 32)``
+    * ``LIBOMPTARGET_REUSE_BLOCKS_FOR_HIGH_TRIP_COUNT=[TRUE/FALSE] (default TRUE)``
```
Comment (jdoerfert) [inline openmp/docs/design/Runtimes.rst:745]: This also needs a section in the docs explaining what it does.

**Review by shiltian — APPROVED** [review]
LGTM

**Pull request merged**